### PR TITLE
feat: median timestamp

### PR DESCRIPTION
**Pull request by joske**
## Motivation

Use the median timestamp of the consensus output to have a deterministic timestamp

## Related PRs
Don't merge until 👇 is merged
https://github.com/AleoHQ/snarkVM/pull/1726

Then don't forget to remove the custom branch for snarkVM in `Cargo.toml`

### CODE REVIEW

File: node/bft-consensus/src/state.rs [hunk 1]
```diff
@@ -130,8 +130,16 @@ impl<N: Network, C: ConsensusStorage<N>> ExecutionState for BftExecutionState<N,
                 return Ok(None);
             }
 
+            // Timestamps are on the batches, not the transactions. We assume every batch contains
+            // at least one valid transaction.
+            let mut timestamps =
+                consensus_output.batches.iter().map(|batch| batch.0.metadata.created_at).collect::<Vec<_>>();
+            timestamps.sort();
+            // why u64, sui???
+            let median: i64 = timestamps[timestamps.len() / 2].try_into().unwrap();
```
Comment (niklaslong) [inline node/bft-consensus/src/state.rs:138]: I don't condone `u64` for timestamps but the rationale is probably that they don't need the neg ones.

File: node/bft-consensus/src/state.rs [hunk 2]
```diff
@@ -130,8 +130,16 @@ impl<N: Network, C: ConsensusStorage<N>> ExecutionState for BftExecutionState<N,
                 return Ok(None);
             }
 
+            // Timestamps are on the batches, not the transactions. We assume every batch contains
+            // at least one valid transaction.
+            let mut timestamps =
+                consensus_output.batches.iter().map(|batch| batch.0.metadata.created_at).collect::<Vec<_>>();
+            timestamps.sort();
+            // why u64, sui???
```
Comment (vvp) [inline node/bft-consensus/src/state.rs]: Probably good to leave this comment out.

File: node/bft-consensus/src/state.rs [hunk 3]
```diff
@@ -130,8 +130,16 @@ impl<N: Network, C: ConsensusStorage<N>> ExecutionState for BftExecutionState<N,
                 return Ok(None);
             }
 
+            // Timestamps are on the batches, not the transactions. We assume every batch contains
+            // at least one valid transaction.
+            let mut timestamps =
+                consensus_output.batches.iter().map(|batch| batch.0.metadata.created_at).collect::<Vec<_>>();
```
Comment (vvp) [inline node/bft-consensus/src/state.rs]: Actually, we need to compute the median timestamp from the certificates in the `sub_dag`.

**Review by niklaslong — COMMENTED** [review]
Not sure what those CI failures are, otherwise lgtm.

File: node/consensus/src/tests.rs [hunk 4]
```diff
@@ -326,7 +327,8 @@ fn test_ledger_deploy() {
     let transactions = consensus.ledger.vm().speculate(sample_finalize_state(1), [transaction.clone()].iter()).unwrap();
 
     // Propose the next block.
-    let next_block = consensus.propose_next_block(&private_key, rng).unwrap();
+    let next_block =
```
Comment (vvp) [inline node/consensus/src/tests.rs]: Add an assertion that the `next_block.timestamp()` matches the given median timestamp.

**Review by vvp — CHANGES_REQUESTED** [review]

File: node/bft-consensus/src/state.rs (hunk 1 above)
Comment (joske) [reply node/bft-consensus/src/state.rs:138]: yes as per the discussion in slack.

**Review by vvp — APPROVED** [review]
LGTM :+1:

**Review by niklaslong — APPROVED** [review]
We'll just need to remember to change the dep after merging the snarkVM PR.

**Review by ljedrz — APPROVED** [review]
So far so good :+1:.

**Comment by joske** [comment]
no longer needed

**Pull request closed without merging**